<div class="alert alert-info">
<h2> Комментарий от ревьюера</h2>
Привет. Меня зовут Алина Гусева, я буду делать ревью к твоей работе по исследоаватльскому анализу данных. В дальнейшем предлагаю общаться на "ты".
    
Видно, что ты старалась, делая проект. Все ключевые этапы проекта выполнены, в целом тебе удалось со всем справиться. Особенно хочется отметить, что соблюдена структура проекта, а код написан аккуратно. 
    
Для того, чтобы и тебе и мне удобнее бы
    ло ориентироваться в комментариях, я буду помечать их разными цветами:
- красный цвет означает комментарии, которые требуют доработки. Как только ты их исправишь - проект будет принят;
- желтый цвет - то, что в следующий раз можно будет сделать по-другому. На твое усмотрение можешь доделать маленькие правки сейчас или учесть замечание для выполнения следующих проектов;
- зеленый цвет - очень элегантные и удачные решение, которые желательно использовать в дальнейшей работе.
- синий цвет - информация
    
Предлагаю работать над проектом в диалоге: если ты что-то изменяешь или вносишь какие-то коррективы - выделяй это каким-нибудь цветом и пиши об этом мне. Так мне гораздо легче будет отследить сделанные изменения.
    
Также попрошу тебя не изменять и не удалять мои комментарии, чтобы проверка твоего самостоятельного проекта прошла оперативнее.
    
 
</div>

## Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

### Шаг 1. Откройте файл с данными и изучите общую информацию. 

Прочитаем файл  и сохраним его в переменной data.

Изучим общую информацию о данных таблицы и получим первые 10 строк таблицы.

In [4]:
import pandas as pd
data = pd.read_csv('/datasets/data.csv')
data.info()
data.head(10)
data.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

### Вывод

Рассмотрим полученную информацию подробнее.

Всего в таблице 12 столбцов, тип данных у столбцов разный: object, int64, float64.

Данные содержат информацию по 21525 клиентам банка. Количество значений в столбцах days_employed и total_income меньше, чем в других. Это говорит о том, что в данных есть пропущенные значения. Более того, столбец days_employed содержит отрицательные значения. Также сразу заметен разный регистр в столбце education.

<div class="alert alert-success">
<h2> Комментарий от ревьюера</h2>
Молодец, хороший тон использовать метод info() для получения общей информации из таблицы. Для более детального изучения можно воспользоваться методом describe()

</div>

### Шаг 2. Предобработка данных

### Обработка пропусков

In [4]:
#обрабатываем пропущенные значения в переменной total_income. Заполним пропущенные значения на средние по группам income_type.
#замена средними по группам income_type будет более точной, чем замена пропущенных значений средним по всем данным 
data['total_income'] = data.groupby('income_type')['total_income'].apply(lambda x: x.fillna(x.mean()))


#обрабатываем пропущенные значения в переменной days_employed. Так этот столбец содержит значения, которые трудно идентифицировать, 
#и он не нужен для ответа на поставленные вопросы, то мы просто можем заполним пропушенные значения 0 и перевести все значения в положительные числа
data['days_employed'] = data['days_employed'].fillna(0)
data['days_employed'] = data['days_employed'].abs()
#проверим преобразования
data.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


### Вывод

Пустые значения свидетельствуют, что для некоторых клиентов банка доступна не вся информация. Причины могут быть, как отсутвие части информации по клиенту, так и проблема с записью данных. 

<div class="alert alert-success">
<h2> Комментарий от ревьюера</h2>
Супер, все верно! Действительно, в столбце total_income много пропущенных значений. Самый логичный способ --- предварительно сгруппировать людей по типу занятости, а замет заменять пропуски. Но воспользоваться лучше все-таки медианным значением, так как оно менее подвержено выбросам.
    
    
Также можно заметить, что пропуски в этих двух столбцах симметричные. Можно придумать способ как заменить пропуски, но поскольку для дальнейшего исселдования на не нужен столбец days_employed, то монжо просто привести его к корректному виду( заменить значения на положительные и избавиться от пропусков)

</div>

### Замена типа данных

In [5]:
#переведем все значения столбцов total_income и days_employed из типа float к целочисленному типу int c помощью метода astype
data['total_income'] = data['total_income'].astype('int')
data['days_employed'] = data['days_employed'].astype('int')
#проверим преобразования
data.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null int64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null int64
purpose             21525 non-null object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


### Вывод

Мы перевели значения столбцов total_income и days_employed к целочисленному типу int, чтобы увеличить точность и избежать неявных ошибок

<div class="alert alert-success">
<h2> Комментарий от ревьюера</h2>
Отлично, теперь все столбцы имеют корректный тип данных.
</div>

### Обработка дубликатов

In [6]:
#посчитаем количество дубликатов 
data.duplicated().sum()
#при взгляде на таблицу сразу бросается в глаза разный регистр в столбце education, проверим
data['education'].value_counts()
#переведем все значения столбца education в нижний регистр
data['education'] = data['education'].str.lower()
#посчитаем количество дубликатов 
data.duplicated().sum()
#также проверим, существует ли проблема разных регистров в столбцах family_status и purpose
data['family_status'].value_counts()
data['purpose'].value_counts()

#обнаружно 71 дубликат, так как и это составляет малую долю от всего объема данных, мы удаляем их 
data = data.drop_duplicates().reset_index(drop = True)
#проверяем
data.duplicated().sum()



0

### Вывод

На этапе предобработки в данных обнаружились не только пропуски, но и дубликаты. Их удаление позволит провести анализ точнее. 

<div class="alert alert-success">
<h2> Комментарий от ревьюера</h2>
Молодец! Здесь важно помнить, что перед обработкой пропусков необходимо привести текстовый формат данных к единому виду, а затем уже избавляться от повторов. Иначе, результат будет иной.

</div>

### Лемматизация

In [7]:
# импортируем pymystem3:
from pymystem3 import Mystem
m = Mystem()

#создаем функцию для лемматизации столбца, для наглядности итоговый результат склеили вызовом метода join()
def lemmas(purpose):
    lemmas_list = ' '.join(m.lemmatize(purpose))
    return lemmas_list

#применяем функцию к стобцу purpose, и записваем результаты в новый столбец lemma_purpose
data['lemma_purpose'] = data['purpose'].apply(lemmas)

#проверяем, что получилось
print(data['lemma_purpose'])

#сразу категоризируем полученные значения, по наиболее часто встречающимся целям кредита
def purpose_category(lemmas):
    if ('жилье' in lemmas) or ('недвижимость' in lemmas):
        return 'Займ на покупку недвижимости'
    if 'образование' in lemmas:
        return 'Займ на образование'
    if 'автомобиль' in lemmas:
        return 'Займ на автомобиль'
    if 'свадьба' in lemmas:
        return 'Займ на свадьбу'
    return 'Другое'
    

#создадим новый столбец category, куда запишем результаты категоризации по целям кредита  
data['category'] = data['lemma_purpose'].apply(purpose_category)
        

0                         покупка   жилье \n
1               приобретение   автомобиль \n
2                         покупка   жилье \n
3            дополнительный   образование \n
4                       сыграть   свадьба \n
                        ...                 
21449                операция   с   жилье \n
21450             сделка   с   автомобиль \n
21451                        недвижимость \n
21452    на   покупка   свой   автомобиль \n
21453           на   покупка   автомобиль \n
Name: lemma_purpose, Length: 21454, dtype: object


In [8]:
#проверим результаты 
print(data['category'].value_counts())

Займ на покупку недвижимости    10811
Займ на автомобиль               4306
Займ на образование              4013
Займ на свадьбу                  2324
Name: category, dtype: int64


### Вывод

Мы лемматизировали столбец purpose, что позволило нам проанализировать текстовые данные и разделить из на категории. Это пригодится нам для ответа на поставленные вопросы. 
Оказалось, что все цели кредита можно разделить на четыре группы: недвижимость, автомобиль, образование и свадьба. Чаще всего люди обращались за кредитами, связанными с недвижимостью, реже всего за кредитами на свадьбу.

<div class="alert alert-success">
<h2> Комментарий от ревьюера</h2>
Ура, лемматизация прошла успешно. Отдельно хочу похвалить за использование функций. Таким образом твой код становится гораздо структурированнее.

</div>

### Категоризация данных

In [9]:
#для ответа на поставленные вопросы необходимо категоризировать данны по количеству детей и по доходу
#начнем с количества детей
data['children'].value_counts()
#в 46 случаях количество детей равно -1, это похоже на опечатку, поэтому переведем все данные в положительные числа 
data['children'] = data['children'].abs()
data['children'].value_counts()
#у 76 заемщиков указано 20 детей, эти данные выглядят нереалистично, поэтому их лучше удалить 
data = data.loc[data['children'] != 20]
print(data['children'].value_counts())

0    14091
1     4855
2     2052
3      330
4       41
5        9
Name: children, dtype: int64


In [10]:
#категоризируем данные по наличию/количеству детей

def family_type(children):
    if children == 0:
        return 'без детей'
    if 1 <= children <= 2:
        return 'малодетная семья'
    if children >= 3:
        return 'многодетная семья'
#создадим новый столбец для категории

data['family_type'] = data['children'].apply(family_type)

In [11]:
#категоризируем данные по доходу 
descriptive = data['total_income'].describe()
quartile_1 = descriptive['25%']
quartile_2 = descriptive['50%']
quartile_3 = descriptive['75%']

def income_category(total_income):
    if total_income <= quartile_1:
        return 'низкий доход'
    if quartile_1 < total_income <= quartile_2:
        return 'средний доход'
    if quartile_2 < total_income <= quartile_3:
        return 'высокий доход'
    if total_income > quartile_3:
        return 'очень высокий доход'

#создадим новый столбец для категории

data['income_category'] = data['total_income'].apply(income_category) 

data.head(10)

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0         340266         53   среднее             1   
5         0            926         27    высшее             0   
6         0           2879         43    высшее             0   
7         0            152         50   среднее             1   
8         2           6929         35    высшее             0   
9         0           2188         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                      purpose                     lemma_purpose  \
0               покупка жилья                покупка   жилье \n   
1     приобретение автомобиля      приобретение   автомобиль \n   
2               покупка жилья                покупка   жилье \n   
3  дополнительное образование   дополнительный   образование \n   
4             сыграть свадьбу              сыграть   свадьба \n   
5               покупка жилья                покупка   жилье \n   
6           операции с жильем           операция   с   жилье \n   
7                 образование                    образование \n   
8       на проведение свадьбы      на   проведение   свадьба \n   
9     покупка жилья для семьи  покупка   жилье   для   семья \n   

                       category        family_type      income_category  
0  Займ на покупку недвижимости   малодетная семья  очень высокий доход  
1            Займ на автомобиль   малодетная семья        средний доход  
2  Займ на покупку недвижимости          без детей        средний доход  
3           Займ на образование  многодетная семья  очень высокий доход  
4               Займ на свадьбу          без детей        высокий доход  
5  Займ на покупку недвижимости          без детей  очень высокий доход  
6  Займ на покупку недвижимости          без детей  очень высокий доход  
7           Займ на образование          без детей        средний доход  
8               Займ на свадьбу   малодетная семья         низкий доход  
9  Займ на покупку недвижимости          без детей        средний доход

### Вывод

Мы также категоризировали данные по количеству детей и доходу.
По количеству детей было выделено три категории: без детей, малодетные семьи (1-2 ребенка) и многодетные семьи (более 3 детей). Во время обработки стобца children, были обнаружены странные значения. Во-первых, отрицательные значения, которые, возможно, были опечаткой. Во-вторых, у 76 человек было указано 20 детей. Так как в данном случаем невозможно установить, откуда появились такие данные, мы их удалили. 
По доходам данные были категоризированы на 4 равные группы по квартилям.


<div class="alert alert-success">
<h2> Комментарий от ревьюера</h2>
Абослютно правильно решение. Во-первых, необходимо избавиться от аномальных значений в столбце children. А во-вторых, очень элегантный и красивый способ разбиения на группы по уровню дохода. Таким образом, получились примерно равные группы, которые позволят провести грамотное исследование.

</div>

### Шаг 3. Ответьте на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [15]:
#создаем сводную табюлицу 
data_pivot = data.pivot_table(index='family_type', columns='debt', values='gender', aggfunc='count')
data_pivot.columns = ['Нет долга', 'Есть долг']
data_pivot['Доля невозратов'] = (data_pivot['Есть долг'] / (data_pivot['Есть долг'] + data_pivot['Нет долга'])).map('{:.1%}'.format)
print(data_pivot)


                   Нет долга  Есть долг Доля невозратов
family_type                                            
без детей              13028       1063            7.5%
малодетная семья        6268        639            9.3%
многодетная семья        349         31            8.2%


### Вывод

Да, зависимость существует. Как видно из сводной таблицы: меньше всего вероятность невозврата кредита у бездетных людей. Чаще других кредит не выплачивают семьи с 1 ребенок или 2 детьми. Что интересно, у многодетных семей вероятность возвращения кредита в срок больше, чем у малодетных, но все-таки меньше, чем у бездетных.  

<div class="alert alert-success">
<h2> Комментарий от ревьюера</h2>
Верно, заемщики без детей самый ответственные.

</div>

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [16]:
family_status_pivot = data.pivot_table(index='family_status', columns='debt', values='gender', aggfunc='count')
family_status_pivot.columns = ['Нет долга', 'Есть долг']
family_status_pivot['Доля невозратов'] = (family_status_pivot['Есть долг'] / (family_status_pivot['Есть долг'] + family_status_pivot['Нет долга'])).map('{:.1%}'.format)
print(family_status_pivot)




                       Нет долга  Есть долг Доля невозратов
family_status                                              
Не женат / не замужем       2528        273            9.7%
в разводе                   1109         84            7.0%
вдовец / вдова               892         63            6.6%
гражданский брак            3754        385            9.3%
женат / замужем            11362        928            7.6%


### Вывод

Чаще всего кредит не возвращают одинокие заемщики (9,7%), в то же время вдовствующие и разведенные заемщики реже всех имеют кредитные долги (6,6% и 7% соответственно). На втором месте по вероятности невозврата кредита находяться люди, состоящие в гражданском браке (9,3%), на третьем - состоящие в официальном браке (7,6%).
Вероятно, здесь также существенную роль может играть возраст заемщиков, так как с одной стороны одинокие люди, которые указали не женат/не замужем хуже всего возвращают кредиты, с другой стороны вдовствующие и разведенные реже всего имеют долги (также можно отметить, что данных по этим двум группам существенно меньше, чем по остальным).

<div class="alert alert-success">
<h2> Комментарий от ревьюера</h2>
Посмотрев на сводную таблицу можно увидеть, что доля просроченых кредитов больше у людей, которые находятся в гражданском браке или не женаты. Чаще всего возращают кредиты вдовцы и вдовы.
</div>


- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [14]:
income_pivot = data.pivot_table(index='income_category', columns='debt', values='gender', aggfunc='count')
income_pivot.columns = ['Нет долга', 'Есть долг']
income_pivot['Доля невозратов'] = (income_pivot['Есть долг'] / (income_pivot['Есть долг'] + income_pivot['Нет долга'])).map('{:.1%}'.format)
print(income_pivot)

                     Нет долга  Есть долг Доля невозратов
income_category                                          
высокий доход             5306        492            8.5%
низкий доход              4918        427            8.0%
очень высокий доход       4547        343            7.0%
средний доход             4874        471            8.8%


### Вывод

Результаты показали, что хуже всего выплачивают кредиты люди со средним доходом, лучше всего - люди с очень высоким доходом.

<div class="alert alert-success">
<h2> Комментарий от ревьюера</h2>
Здесь очент сильно влияет то, на какие группы ты разбила. Поскольку выборки по объему одинаковые, то можно сделать определенные выводы. Один из них, что люди с средним доходм имеют больше всего задолжностей.
</div>


- Как разные цели кредита влияют на его возврат в срок?

In [79]:
purpose_pivot = data.pivot_table(index='category', columns='debt', values='gender', aggfunc='count')
purpose_pivot.columns = ['Нет долга', 'Есть долг']
purpose_pivot['Доля невозратов'] = (purpose_pivot['Есть долг'] / (purpose_pivot['Есть долг'] + purpose_pivot['Нет долга'])).map('{:.1%}'.format)
print(purpose_pivot)

                              Нет долга  Есть долг Доля невозратов
category                                                          
Займ на автомобиль                 3898        401            9.3%
Займ на образование                3638        369            9.2%
Займ на покупку недвижимости      10024        780            7.2%
Займ на свадьбу                    2156        183            7.8%


### Вывод

Чаще всего не возвращали кредиты на покупку автомобиля, реже всего на покупку недвижимости.

### Шаг 4. Общий вывод

Как показали результаты исследования, наличие детей, семейное положение, уровень дохода и цель кредита влияют на вероятность его возврата в срок. С наибольшей вероятность кредит вернет резведенный или вдовствующий заемщик без детей с очень высоким дохом, который берет кредит на покупку недвижимости. С наименьшей вероятностью - не женатый/не замужний заемщик с 1-2 детьми со средним доходом, который берет кредит на покупку автомобиля. Однако стоит отметить, что данных по семьям с 3 и более детьми в данных таблицы не очень много (всего 349), поэтому для более точных прогнозов необходимо больше данных по многодетным семьям.

<div class="alert alert-success">
<h2> Комментарий от ревьюера</h2>

Поздравляю с первым успешным проектом!

- Ты хорошо делаешь выводы и анализируешь результаты.
    
- Код написан аккуратно и качественно, видно, что есть какой-то определенный бэкграунд.

- Соблюдена структура проекта, было приятно проверять:)
    
- Все написано четко и по делу

    
Желаю дальнейших успехов!


</div>


### Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение, какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено, по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.